### PR TITLE
Support default value when create tag and edge. 

### PR DESCRIPTION
**Pull request by darionyaphet**
#631

### CODE REVIEW

File: src/graph/InsertEdgeExecutor.cpp [hunk 1]
```diff
@@ -125,6 +137,78 @@ StatusOr<std::vector<storage::cpp2::Edge>> InsertEdgeExecutor::prepareEdges() {
             values.emplace_back(ovalue.value());
         }
 
+        if (expressions.size() != schema_->getNumFields()) {
```
Comment (wadeliuyi) [inline src/graph/InsertEdgeExecutor.cpp]: this function is too long, maybe should separate it to some sub function?

File: src/graph/InsertEdgeExecutor.cpp [hunk 2]
```diff
@@ -125,6 +137,78 @@ StatusOr<std::vector<storage::cpp2::Edge>> InsertEdgeExecutor::prepareEdges() {
...
+                                variantType = folly::to<bool>(defaultValues_[fieldName]);
+                            } catch (const std::exception& ex) {
+                                LOG(ERROR) << "Conversion to bool failed: "
+                                           << defaultValues_[fieldName];
+                                return Status::Error("Type Conversion Failed");
+                            }
+
+                            VLOG(3) << "Insert " << fieldName << ":" << variantType
+                                    << " at " << insertPosition;
+                            values.insert(values.begin() + insertPosition, std::move(variantType));
```
Comment (wadeliuyi) [inline src/graph/InsertEdgeExecutor.cpp]: in this function, maybe list for values is better than vector.

File: src/graph/InsertEdgeExecutor.cpp [hunk 3]
```diff
@@ -125,6 +137,78 @@ StatusOr<std::vector<storage::cpp2::Edge>> InsertEdgeExecutor::prepareEdges() {
...
+                            break;
+                        default:
+                            LOG(ERROR) << "Unknow type";
+                            return Status::Error("Unknow type");
+                    }
+                    insertPosition++;
+                } else if (*props_[propsIndex] == fieldName) {
```
Comment (wadeliuyi) [inline src/graph/InsertEdgeExecutor.cpp]: why here we can visit props_ by order, but in function check we visit it by find_if ?

File: src/graph/InsertVertexExecutor.cpp [hunk 4]
```diff
@@ -111,22 +135,83 @@ StatusOr<std::vector<storage::cpp2::Vertex>> InsertVertexExecutor::prepareVertic
...
             auto schema = schemas_[index];
 
-            // props's number should equal to value's number
-            if ((valuePos + propSize) > values.size()) {
-                LOG(ERROR) << "Input values number " << values.size()
-                           << ", props number " << propSize;
-                return Status::Error("Wrong number of value");
+            auto schemaNumFields = schema->getNumFields();
+            size_t propsIndex = 0;
+            for (size_t schemaIndex = 0; schemaIndex < schemaNumFields; schemaIndex++) {
```
Comment (wadeliuyi) [inline src/graph/InsertVertexExecutor.cpp]: maybe can do this in sub function

File: src/graph/InsertVertexExecutor.cpp [hunk 5]
```diff
@@ -111,22 +135,83 @@ StatusOr<std::vector<storage::cpp2::Vertex>> InsertVertexExecutor::prepareVertic
...
+                        case nebula::cpp2::SupportedType::STRING:
+                            variantType = defaultValues_[fieldName];
+                            values.insert(values.begin() + insertPosition, std::move(variantType));
+                            break;
+                        default:
+                            LOG(ERROR) << "Unknow type";
+                            return Status::Error("Unknow type");
+                    }
+                    insertPosition++;
+                } else if (*props[propsIndex] == fieldName) {
```
Comment (wadeliuyi) [inline src/graph/InsertVertexExecutor.cpp]: else if (propsIndex < props.size()) {
    if (*props[propsIndex] == fieldName) {
        propsIndex++;
    }
    insertPosition++;
}

File: src/graph/InsertVertexExecutor.cpp [hunk 6]
```diff
@@ -111,22 +135,83 @@ StatusOr<std::vector<storage::cpp2::Vertex>> InsertVertexExecutor::prepareVertic
...
+                    insertPosition++;
+                } else {
+                    insertPosition++;
+                }
             }
 
             RowWriter writer(schema);
             auto valueIndex = valuePos;
-            for (auto fieldIndex = 0u; fieldIndex < schema->getNumFields(); fieldIndex++) {
+            for (auto fieldIndex = 0u; fieldIndex < schemaNumFields; fieldIndex++) {
```
Comment (wadeliuyi) [inline src/graph/InsertVertexExecutor.cpp]: no auto when define a variable

File: src/meta/client/MetaClient.cpp [hunk 7]
```diff
@@ -1140,6 +1140,50 @@ folly::Future<StatusOr<bool>> MetaClient::balanceLeader() {
     return future;
 }
 
+folly::Future<StatusOr<std::string>> MetaClient::getTagDefaultValue(GraphSpaceID spaceId,
+                                                                    TagID tagId,
+                                                                    const std::string& field) {
+    cpp2::GetReq req;
+    static std::string defaultKey = "__default__";
```
Comment (wadeliuyi) [inline src/meta/client/MetaClient.cpp]: the default value maybe can cached in client.

File: src/meta/processors/schemaMan/CreateEdgeProcessor.cpp [hunk 8]
```diff
@@ -49,7 +49,65 @@ void CreateEdgeProcessor::process(const cpp2::CreateEdgeReq& req) {
     data.emplace_back(MetaServiceUtils::schemaEdgeKey(req.get_space_id(), edgeType, 0),
                       MetaServiceUtils::schemaEdgeVal(req.get_edge_name(), req.get_schema()));
 
-    LOG(INFO) << "Create Edge  " << req.get_edge_name() << ", edgeType " << edgeType;
+    auto columns = req.get_schema().get_columns();
+    for (auto& column : columns) {
+        if (column.__isset.default_value) {
```
Comment (wadeliuyi) [inline src/meta/processors/schemaMan/CreateEdgeProcessor.cpp]: do it in sub function

File: src/graph/InsertEdgeExecutor.cpp (hunk 1 above)
Comment (darionyaphet) [reply src/graph/InsertEdgeExecutor.cpp]: I have extracted some common parts.

File: src/meta/client/MetaClient.cpp (hunk 7 above)
Comment (darionyaphet) [reply src/meta/client/MetaClient.cpp]: Yep Good Point ~

File: src/graph/InsertEdgeExecutor.cpp [hunk 9]
```diff
@@ -125,6 +137,78 @@ StatusOr<std::vector<storage::cpp2::Edge>> InsertEdgeExecutor::prepareEdges() {
...
+                    propsIndex++;
+                    insertPosition++;
+                } else {
+                    insertPosition++;
```
Comment (critical27) [inline src/graph/InsertEdgeExecutor.cpp]: Why we can increase the position here? It seems you handle the situation when `props_[propsIndex]` match with `fieldName` in if statement, and mismatch in previous else if. Forgive me, a little bit hard to comprehend.

**Comment by whitewum** [comment]
Question: How to modify the default value?  for the first time, I set int as 0; and the second time, I hope a int is -1?

**Comment by darionyaphet** [comment]
> Question: How to modify the default value? for the first time, I set int as 0; and the second time, I hope a int is -1?

According to `MySQL`, it could use `ALTER TABLE` to modify the default value's data type and value. 

Maybe there are some related works  for this PR for example: `show create table with default info` and `alter default value`

File: src/graph/InsertEdgeExecutor.cpp [hunk 10]
```diff
@@ -150,31 +147,84 @@ StatusOr<std::vector<storage::cpp2::Edge>> InsertEdgeExecutor::prepareEdges() {
...
+                        break;
+                    case nebula::cpp2::SupportedType::STRING:
+                        value = defaultValues_[fieldName];
+                        break;
+                    default:
+                        LOG(ERROR) << "Unknow type";
+                        return Status::Error("Unknow type");
+                }
+                VLOG(3) << "Supplement default value : " << fieldName << " : " << value;
+                if (schemaType.type == nebula::cpp2::SupportedType::TIMESTAMP) {
```
Comment (critical27) [inline src/graph/InsertEdgeExecutor.cpp]: How about reuse same code with Line 166

File: src/graph/InsertVertexExecutor.cpp [hunk 11]
```diff
@@ -124,55 +142,101 @@ StatusOr<std::vector<storage::cpp2::Vertex>> InsertVertexExecutor::prepareVertic
...
                     }
-                    writeVariantType(writer, timestamp.value());
+
                 } else {
-                    writeVariantType(writer, value);
-                }
+                    // fetch default value from cache
+                    VariantType value;
+                    auto type = schema->getFieldType(schemaIndex).type;
+                    switch (type) {
```
Comment (critical27) [inline src/graph/InsertVertexExecutor.cpp]: It seems this piece is same with InsertEdge, could we reuse them?

Comment (darionyaphet) [reply src/graph/InsertVertexExecutor.cpp]: Thanks & I will move it into `Executor`

**Pull request merged**